datafusion/expr/src/logical_plan/mutate.rs

```diff
@@ -0,0 +1,346 @@
+            Expr::Exists(Exists { subquery, .. })
+            | Expr::InSubquery(InSubquery { subquery, .. })
+            | Expr::ScalarSubquery(subquery) => {
+                rewrite_arc(&mut subquery.subquery, &mut f)
+            }
+            _ => Ok(Transformed::no(())),
+        })
+    }
+}
+
+/// writes each `&mut Expr` in the iterator using `f`
+fn rewrite_expr_iter_mut<'a, F>(
+    i: impl IntoIterator<Item = &'a mut Expr>,
+    mut f: F,
+) -> Result<Transformed<()>>
+where
+    F: FnMut(&mut Expr) -> Result<Transformed<()>>,
+{
+    i.into_iter()
+        .try_fold(Transformed::no(()), |acc, expr| acc.and_then(|| f(expr)))
+}
+
+/// A temporary node that is left in place while rewriting the children of a
+/// [`LogicalPlan`]. This is necessary to ensure that the `LogicalPlan` is
+/// always in a valid state (from the Rust perspective)
+static PLACEHOLDER: OnceLock<Arc<LogicalPlan>> = OnceLock::new();
+
+/// Applies `f` to rewrite the existing node, while avoiding `clone`'ing as much
+/// as possiblw.
+///
+/// TODO eventually remove `Arc<LogicalPlan>` from `LogicalPlan` and have it own
+/// its inputs, so this code would not be needed. However, for now we try and
+/// unwrap the `Arc` which avoids `clone`ing in most cases.
+///
+/// On error, node be left with a placeholder logical plan
+fn rewrite_arc<F>(node: &mut Arc<LogicalPlan>, mut f: F) -> Result<Transformed<()>>
+where
+    F: FnMut(&mut LogicalPlan) -> Result<Transformed<()>>,
+{
+    // We need to leave a valid node in the Arc, while we rewrite the existing
+    // one, so use a single global static placeholder node
+    let mut new_node = PLACEHOLDER
+        .get_or_init(|| {
+            Arc::new(LogicalPlan::EmptyRelation(EmptyRelation {
+                produce_one_row: false,
+                schema: DFSchemaRef::new(DFSchema::empty()),
+            }))
+        })
+        .clone();
+
+    // take the old value out of the Arc
+    std::mem::swap(node, &mut new_node);
+
+    // try to update existing node, if it isn't shared with others
+    let mut new_node = Arc::try_unwrap(new_node)
+        // if None is returned, there is another reference to this
+        // LogicalPlan, so we must clone instead
+        .unwrap_or_else(|node| node.as_ref().clone());
+
+    // apply the actual transform
+    let result = f(&mut new_node)?;
+
+    // put the new value back into the Arc
+    let mut new_node = Arc::new(new_node);
+    std::mem::swap(node, &mut new_node);
+
+    Ok(result)
+}
+
+/// Rewrites a [`Column`] in place using the provided closure
+fn rewrite_column<F>(column: &mut Column, mut f: F) -> Result<Transformed<()>>
+where
+    F: FnMut(&mut Expr) -> Result<Transformed<()>>,
+{
+    // Since `Column`'s isn't an `Expr`, but the closure in terms of Exprs,
+    // we make a temporary Expr to rewrite and then put it back
+
+    let mut swap_column = Column::new_unqualified("TEMP_unnest_column");
+    std::mem::swap(column, &mut swap_column);
```

Comment:
I think we can have mem::take here too with a default implementation for Column

Reply:
I tried using `take` and it seems like there is no default impl for `Column` 🤔 

```
error[E0277]: the trait bound `datafusion_common::Column: std::default::Default` is not satisfied
   --> datafusion/expr/src/logical_plan/mutate.rs:288:41
    |
288 |     let mut swap_column= std::mem::take(column);
    |                          -------------- ^^^^^^ the trait `std::default::Default` is not implemented for `datafusion_common::Column`
    |                          |
    |                          required by a bound introduced by this call
```

Reply:
yes, I would prefer "" as default column if it makes sense

Reply:
I see -- the reason I left a `TEMP...` name here was so that on error it was obvious that the Column was partially rewritten.

Given the comments below, however, it seems like we may not even need this API